cub/warp/warp_exchange.cuh

```diff
@@ -468,5 +468,89 @@ public:
   //@}  end member group
 };
 
+template <typename InputT, int ITEMS_PER_THREAD>
+class WarpExchangeRegister
+{
+  static_assert(PowerOfTwo<ITEMS_PER_THREAD>::VALUE, "ITEMS_PER_THREAD must be a power of two");
+
+  template <int SIZE, int idx>
+  class CompileTimeArrayImpl; // fwd decl for partial specializations
+
+  template <int SIZE> // terminating partial specialization
+  class CompileTimeArrayImpl<SIZE, SIZE>
+  {
+  protected:
+    static_assert(PowerOfTwo<SIZE>::VALUE, "SIZE must be a power of two");
+
+    // used for dumping back the individual values after transposing
+    InputT (&output_items)[SIZE];
+
+    template <int>
+    __device__ void Loop(bool, unsigned)
+    {}
+
+  public:
+    __device__ CompileTimeArrayImpl(InputT (&input_items)[SIZE])
+        : output_items{input_items}
+    {}
+  };
+
+  template <int SIZE, int idx> // concrete recursion class
+  class CompileTimeArrayImpl : protected CompileTimeArrayImpl<SIZE, idx + 1>
+  {
+  protected:
+    static_assert(PowerOfTwo<SIZE>::VALUE, "SIZE must be a power of two");
+
+    InputT val;
+
+    template <int NUM_ENTRIES>
+    __device__ void Loop(const bool xor_bit_set, const unsigned mask)
+    {
+      // NOTE: Do *NOT* try to refactor this code to use a reference, since nvcc
+      //       tends to choke on it and then drop everything into local memory.
+      const InputT send_val = (xor_bit_set ? CompileTimeArrayImpl<SIZE, idx>::val
+                                           : CompileTimeArrayImpl<SIZE, idx + NUM_ENTRIES>::val);
+      const InputT recv_val = __shfl_xor_sync(mask, send_val, NUM_ENTRIES, SIZE);
```

Comment:
It's optional, but it would be nice if you could document the algorithm logic here.

Reply:
I will document everything once the code is in shape (since I don't want to be editing the Doxygen repeatedly), especially since the underlying idea isn't evident to people that haven't thought a lot about this problem.

Reply:
I don't anticipate you providing Doxygen documentation as part of this PR. The API is documented for `WarpExchange`. The shuffle algorithm is going to be an implementation detail, so it doesn't have to document API. What I meant is providing some comments regarding the algorithm logic.

Reply:
implementation documented